plugins/src/gcc/imp.rs

```diff
@@ -0,0 +1,1270 @@
+
+    last_received_packets: VecDeque<Packet>,
+
+    // Last loss update
+    last_loss_update: Option<time::Instant>,
+    // Moving average of the packet loss
+    loss_average: f64,
+
+    // Kalman filter fields
+    gain: f64,
+    measurement_uncertainty: f64, // var_v_hat(i-1)
+    estimate_error: f64,          // e(i-1)
+    estimate: Duration,           // m_hat(i-1)
+
+    // Threshold fields
+    threshold: Duration,
+    last_threshold_update: Option<time::Instant>,
+    num_deltas: i64,
+
+    // Overuse related fields
+    increasing_counter: u32,
+    last_overuse_estimate: Duration,
+    last_use_detector_update: time::Instant,
+    increasing_duration: Duration,
+
+    // round-trip-time estimations
+    rtts: VecDeque<Duration>,
+    clock: gst::Clock,
+
+    // Current network usage state
+    usage: NetworkUsage,
+}
+
+// Monitors packet loss and network overuse through because of delay
+impl Debug for Detector {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "Network Usage: {:?}. Effective bitrate: {}ps - Measure: {} Estimate: {} threshold {} - overuse_estimate {}",
+            self.usage,
+            human_kbits(self.effective_bitrate()),
+            pdur(&self.measure),
+            pdur(&self.estimate),
+            pdur(&self.threshold),
+            pdur(&self.last_overuse_estimate),
+        )
+    }
+}
+
+impl Detector {
+    fn new() -> Self {
+        Detector {
+            group: PacketGroup::new(),
+            prev_group: Default::default(),
+            measure: Duration::zero(),
+            last_inter_departure_times: Vec::with_capacity(LAST_INTER_DEPARTURE_SIZE + 1),
+
+            /* Smallish value to hold PACKETS_RECEIVED_WINDOW packets */
+            last_received_packets: VecDeque::with_capacity(400),
+
+            last_loss_update: None,
+            loss_average: 0.,
+
+            gain: 0.,
+            measurement_uncertainty: 0.,
+            estimate_error: INITIAL_ERROR_COVARIANCE,
+            estimate: Duration::zero(),
+
+            threshold: *INITIAL_DEL_VAR_TH,
+            last_threshold_update: None,
+            num_deltas: 0,
+
+            last_use_detector_update: time::Instant::now(),
+            increasing_counter: 0,
+            last_overuse_estimate: Duration::zero(),
+            increasing_duration: Duration::zero(),
+
+            rtts: Default::default(),
+            clock: gst::SystemClock::obtain(),
+
+            usage: NetworkUsage::Normal,
+        }
+    }
+
+    fn loss_ratio(&self) -> f64 {
+        self.loss_average
+    }
+
+    fn update_last_received_packets(&mut self, packet: Packet) {
+        let arrival = packet.arrival;
+        let idx = match self
+            .last_received_packets
+            .binary_search_by(|packet| packet.arrival.cmp(&arrival))
+        {
+            Ok(idx) => idx,
+            Err(idx) => idx,
+        };
+        self.last_received_packets.insert(idx, packet);
```

Comment:
Instead of a `VecDeque` a `BTreeMap<usize, _>` might be more efficient?

Reply:
Using a `VecDeque` here does seem not very adapted (given the requirement for "random" insertion, and the systematic binary search), I don't understand your comment about the seqnum @thiblahute ?

Reply:
You work with "extended seqnums" for that reason usually (roll-over counter in the upper, extended bits)

Reply:
Would need to bring that info into the TWCC stats event/structure I guess. could also probably find a way to handle the rollover counter myself.

Reply:
Implemented.